### PR TITLE
feat: Firebase Segment

### PR DESCRIPTION
**Pull request by ivan-the-terrible**
### Prerequisites

- [x] I have read and understood the [contributing guide][CONTRIBUTING.md].
- [x] The commit message follows the [conventional commits][cc] guidelines.
- [x] Tests for the changes have been added (for bug fixes / features).
- [x] Docs have been added/updated (for bug fixes / features).

### Description

After discovering a user who created a custom script to show the current Firebase project, the need for a real Firebase segment came to light.
Here is the original script invoking the idea: https://github.com/QuentinPhrl/Omp_firebase_prompt

The segment, its tests, its required references like in schema and in the engine, and the proper documentation have all been added.

Firebase Segment

### Firebase Segment

In short, here is how the segment works:

Once a user has installed the `firebase-tools` CLI npm package, they can initialize and connect a Firebase project.

The CLI tool creates `~/.config/configstore/firebase-tools.json` file and any directory that has had `firebase init` ran on it and a project associated with it, gets added to the JSON file via the `activeProjects` property. This property has the path as the key and the project name as the value.

image

If the Firebase segment is active, we are checking that `activeProjects` map to see if the curent directory exists in there. If it does, we display the project name in the segment.

Segment Doc Page:

Firebase Documentation

[CONTRIBUTING.md]: https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/CONTRIBUTING.md
[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

### CODE REVIEW

File: src/segments/firebase.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	return " {{ .Project}} "
+}
+
+func (f *Firebase) Init(props properties.Properties, env platform.Environment) {
+	f.props = props
+	f.env = env
+}
+
+func (f *Firebase) Enabled() bool {
+	cfgDir := path.Join(f.env.Home(), ".config", "configstore")
```
Comment (JanDeDobbeleer) [inline src/segments/firebase.go]:
```suggestion
	cfgDir := filepath.Join(f.env.Home(), ".config", "configstore")
```

File: src/segments/firebase.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	configFile, err := f.getActiveConfig(cfgDir)
+	if err != nil {
+		f.env.Error(err)
+		return false
+	}
+	data, err := f.getFirebaseData(configFile)
```
Comment (JanDeDobbeleer) [inline src/segments/firebase.go:44]: add a space for readability between these blocks

File: src/segments/firebase.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+	// Test if the current directory is a project path
+	// and if it is, return the project name
+	for key, value := range data.ActiveProject {
+		if key == f.env.Pwd() {
+			f.Project = value
+			return true
+		}
+	}
+	return false
```
Comment (JanDeDobbeleer) [inline src/segments/firebase.go:62]: add a space for readability before `return false`

File: src/segments/firebase.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func (f *Firebase) getActiveConfig(cfgDir string) (string, error) {
+	activeConfigFile := path.Join(cfgDir, "firebase-tools.json")
```
Comment (JanDeDobbeleer) [inline src/segments/firebase.go]:
```suggestion
	activeConfigFile := filepath.Join(cfgDir, "firebase-tools.json")
```

File: src/segments/firebase.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+const (
+	FIREBASENOACTIVECONFIG = "NO ACTIVE CONFIG FOUND"
+)
+
+type Firebase struct {
+	props properties.Properties
+	env   platform.Environment
+
+	Project string
+	FirebaseData
```
Comment (JanDeDobbeleer) [inline src/segments/firebase.go]: We shouldn't expose this, right?

File: src/segments/firebase.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	activeConfigData := f.env.FileContent(activeConfigFile)
+	if len(activeConfigData) == 0 {
+		return "", errors.New(FIREBASENOACTIVECONFIG)
+	}
+	return activeConfigData, nil
+}
+
+func (f *Firebase) getFirebaseData(configFile string) (*FirebaseData, error) {
+	var data FirebaseData
+	err := json.Unmarshal([]byte(configFile), &data)
```
Comment (JanDeDobbeleer) [inline src/segments/firebase.go:77]:
```suggestion

	err := json.Unmarshal([]byte(configFile), &data)
```

File: src/segments/firebase.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &data, nil
```
Comment (JanDeDobbeleer) [inline src/segments/firebase.go:82]:
```suggestion

	return &data, nil
```

**Review by JanDeDobbeleer — CHANGES_REQUESTED** [review]

File: src/segments/firebase.go (hunk 3 above)
Comment (ivan-the-terrible) [reply src/segments/firebase.go:62]: Done

File: src/segments/firebase.go (hunk 1 above)
Comment (ivan-the-terrible) [reply src/segments/firebase.go]: Updated to `filepath`

File: src/segments/firebase.go (hunk 2 above)
Comment (ivan-the-terrible) [reply src/segments/firebase.go:44]: Space added

File: src/segments/firebase.go (hunk 4 above)
Comment (ivan-the-terrible) [reply src/segments/firebase.go]: Updated to `filepath`

File: src/segments/firebase.go (hunk 7 above)
Comment (ivan-the-terrible) [reply src/segments/firebase.go:82]: Added space

File: src/segments/firebase.go (hunk 6 above)
Comment (ivan-the-terrible) [reply src/segments/firebase.go:77]: added space

File: src/segments/firebase.go (hunk 5 above)
Comment (ivan-the-terrible) [reply src/segments/firebase.go]: Totally right, it's unnecessary to expose it. I think I grabbed that idea from another segment.
Nonetheless, it's been removed

**Comment by JanDeDobbeleer** [comment]
@all-contributors please add @ivan-the-terrible for code

**Pull request merged**